src/main/java/runhistoryplus/patches/ImprovableCardRunHistoryPatch.java

```diff
@@ -0,0 +1,88 @@
+package runhistoryplus.patches;
+
+import basemod.ReflectionHacks;
+import com.evacipated.cardcrawl.modthespire.lib.*;
+import com.evacipated.cardcrawl.modthespire.patcher.PatchingException;
+import com.megacrit.cardcrawl.cards.AbstractCard;
+import com.megacrit.cardcrawl.cards.CardGroup;
+import com.megacrit.cardcrawl.cards.blue.GeneticAlgorithm;
+import com.megacrit.cardcrawl.cards.colorless.RitualDagger;
+import com.megacrit.cardcrawl.core.CardCrawlGame;
+import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
+import com.megacrit.cardcrawl.metrics.Metrics;
+import com.megacrit.cardcrawl.monsters.MonsterGroup;
+import com.megacrit.cardcrawl.screens.runHistory.RunHistoryScreen;
+import com.megacrit.cardcrawl.screens.runHistory.TinyCard;
+import com.megacrit.cardcrawl.screens.stats.RunData;
+import javassist.CannotCompileException;
+import javassist.NotFoundException;
+import javassist.CtBehavior;
+import javassist.CtClass;
+import javassist.CtField;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ImprovableCardRunHistoryPatch {
+    private static Map<String, List<Integer>> improvableCardsLog;
+    private static final ArrayList<String> improvableCards = new ArrayList<String>() {{
+        add(GeneticAlgorithm.ID);
+        add(RitualDagger.ID);
+    }};
+
+    @SpirePatch(clz = CardCrawlGame.class, method = SpirePatch.CONSTRUCTOR)
+    public static class ImprovableCardsField {
+        @SpireRawPatch
+        public static void addImprovableCardsField(CtBehavior ctBehavior) throws NotFoundException, CannotCompileException {
+            CtClass runData = ctBehavior.getDeclaringClass().getClassPool().get(RunData.class.getName());
+            String fieldSource = "public java.util.Map improvable_cards;";
+            CtField field = CtField.make(fieldSource, runData);
+            runData.addField(field);
+        }
+    }
+
+    @SpirePatch(clz = Metrics.class, method = "gatherAllData")
+    public static class GatherAllDataPatch {
+        @SpirePostfixPatch
+        public static void gatherAllDataPatch(Metrics __instance, boolean death, boolean trueVictor, MonsterGroup monsters) {
+            improvableCardsLog = new HashMap<>();
+            for (final AbstractCard card: AbstractDungeon.player.masterDeck.group) {
+                if (improvableCards.contains(card.cardID)) {
+                    if (!improvableCardsLog.containsKey(card.name)) {
```

Comment:
Is the intent of recording entries by `card.name` here to distinguish between upgraded and unupgraded versions of cards (because one will be "Genetic Algorithm", while the other is "Genetic Algorithm+"?

If so, there's a different way to do this that should work better, which is to use `card.getMetricID()` instead of `card.name`. That's how other code related to run history distinguishes between upgraded and unupgraded versions.

The downside of the current approach that the name will be language-specific, which means that (1) if you do a run in one language and then look at it in another language, this won't work and (2) if you're using these files to analyze runs, the entry you look for will depend on the language. That's why I recommend switching everything to use `card.getMetricID()`.

Reply:
Discriminating between (un)-upgraded was intentional, because there are separate entries for both in the deck list on the run history screen. I saw `getMetricID` but didn't use it, because it slightly simplified the matching at some point, but I wasn't aware of the localization issue, so I will definitely change that!